File: src/lib/leaderboard.ts
/**
 * Chicken Leaderboard
 * Tracking poultry consumption across the enterprise
 */

interface LeaderboardEntry {
	clientId: string;
	totalChickens: number;
	lastRequest: Date;
}

// In-memory store (resets on server restart, as intended)
const leaderboard: Map<string, LeaderboardEntry> = new Map();

export function recordChickenRequest(clientId: string, count: number): void {
	const existing = leaderboard.get(clientId);
	if (existing) {
		existing.totalChickens += count;
		existing.lastRequest = new Date();
	} else {
		leaderboard.set(clientId, {
			clientId,
			totalChickens: count,
			lastRequest: new Date()
		});
	}
}

export function getLeaderboard(): LeaderboardEntry[] {
	return Array.from(leaderboard.values())
		.sort((a, b) => b.totalChickens - a.totalChickens)
		.slice(0, 10); // Top 10 chicken consumers
}

export function getClientStats(clientId: string): LeaderboardEntry | undefined {
	return leaderboard.get(clientId);
}

File: backend/app/core/leaderboard.py
"""
In-memory leaderboard for tracking chicken consumption
"""

from collections import defaultdict

# In-memory storage (resets on server restart)
_leaderboard: dict[str, int] = defaultdict(int)


def record_request(client_id: str, chicken_count: int) -> None:
    """Record a chicken request for a client"""
    _leaderboard[client_id] += chicken_count


def get_top_consumers(limit: int = 10) -> list[dict]:
    """Get the top chicken consumers"""
    sorted_consumers = sorted(
        _leaderboard.items(),
        key=lambda x: x[1],
        reverse=True
    )[:limit]

    return [
        {"client": client, "chickens": count, "rank": idx + 1}
        for idx, (client, count) in enumerate(sorted_consumers)
    ]


def reset_leaderboard() -> None:
    """Reset the leaderboard (for testing)"""
    _leaderboard.clear()
